abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,187 @@
+            )
+            containers.append(container)
+
+    # wait for all container runs to complete (status = "exited")
+    wait_for_containers(containers, timeout)
+    logger.info(f"All {len(containers)} containers have exited.")
+
+    # get logs from all Docker containers
+    log_file = f"{run_directory}/transformed/logs.txt"
+    with open(Path(log_file), "w") as file:
+        for container in containers:
+            header = (
+                f"docker_image: {container.labels['docker_image']} | "
+                f"source: {container.labels['source']} | "
+                f"input_file: {container.labels['input_file']}\n"
+            )
+            file.write(header)
+            file.write(container.logs().decode())
+            file.write("\n\n")
+    logger.info(f"Log file created: {log_file}")
+
+    # update run.json with list of input files
+    update_or_create_run_json(run_directory, new_data={"input_files": input_files})
+
+    elapsed_time = perf_counter() - start_time
+    logger.info(
+        "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
+    )
+    return get_transformed_files(run_directory)
+
+
+def run_docker_container(
+    docker_image: str,
+    transformed_directory: str,
+    input_file: str,
+    docker_client: docker.client.DockerClient,
+) -> Container:
+    """Run transmogrifier via Docker container to transform input file."""
+    source, output_filename = parse_transform_details_from_extract_filename(input_file)
+    container = docker_client.containers.run(
+        docker_image,
+        command=[
+            f"--input-file={input_file}",
+            f"--output-file=/tmp/{output_filename}",
+            f"--source={source}",
+        ],
+        detach=True,
+        environment={
+            "AWS_ACCESS_KEY_ID": CONFIG.AWS_ACCESS_KEY_ID,
+            "AWS_SECRET_ACCESS_KEY": CONFIG.AWS_SECRET_ACCESS_KEY,
+            "AWS_SESSION_TOKEN": CONFIG.AWS_SESSION_TOKEN,
+        },
+        labels={
+            "docker_image": docker_image,
+            "source": source,
+            "input_file": input_file,
+        },
+        volumes=[f"{transformed_directory}:/tmp"],
+    )
+    logger.info(
+        f"Container '{container.id}' (Docker image: {docker_image}) "
+        f"RUNNING transform for '{source}' input_file: {input_file}."
+    )
+    return container
+
+
+def wait_for_containers(containers: list[Container], timeout: int = 180) -> None:
+    """Given a list of running containers, wait until all containers exit."""
+    exited_containers: list[Container] = []
+    start_time = time.time()
+    while len(exited_containers) < len(containers):
+        running_containers = [
+            container for container in containers if container not in exited_containers
+        ]
+        if time.time() - start_time >= timeout:
+            raise DockerContainerRuntimeExceededTimeoutError(
+                containers=containers, timeout=timeout
+            )
```

Comment:
Nice addition!  I like this, if I'm understanding correctly, as it will report _all_ containers that have exceeded the timeout.

Reply:
Yes, it will -- once I update `containers=containers` -> `containers=running_containers` 😅 . Change has been made.

Reply:
Hahaha, I didn't even catch that!  Your change is how I had read the code.  Nice catch (agree that we only want to print out the timed-out containers).